src/lib/homestore_backend/snapshot_receive_handler.cpp

```diff
@@ -0,0 +1,213 @@
+            if (std::memcmp(computed_hash, header->hash, BlobHeader::blob_max_hash_len) != 0) {
+                LOGE("Hash mismatch for blob_id {}: header [{}] [computed={:np}]", blob->blob_id(), header->to_string(),
+                     spdlog::to_hex(computed_hash, computed_hash + BlobHeader::blob_max_hash_len));
+                return BLOB_DATA_CORRUPTED;
+            }
+        }
+
+        // Alloc & persist blob data
+        auto chunk_id = home_obj_.get_shard_chunk(shard_cursor_);
+        RELEASE_ASSERT(chunk_id.has_value(), "Failed to load chunk of current shard_cursor:{}", shard_cursor_);
+        homestore::blk_alloc_hints hints;
+        hints.chunk_id_hint = *chunk_id;
+
+        auto data_size = blob->data()->size();
+        homestore::MultiBlkId blk_id;
+        auto status = homestore::data_service().alloc_blks(
+            sisl::round_up(data_size, homestore::data_service().get_blk_size()), hints, blk_id);
+        if (status != homestore::BlkAllocStatus::SUCCESS) {
+            LOGE("Failed to allocate blocks for shard {} blob {}", shard_cursor_, blob->blob_id());
+            return ALLOC_BLK_ERR;
+        }
+
+        auto free_allocated_blks = [blk_id]() {
+            homestore::data_service().async_free_blk(blk_id).thenValue([blk_id](auto&& err) {
+                LOGD("Freed blk_id:{} due to failure in adding blob info, err {}", blk_id.to_string(),
+                     err ? err.message() : "nil");
+            });
+        };
+
+        const auto ret = homestore::data_service()
+                             .async_write(r_cast< char const* >(blob_data), data_size, blk_id)
+                             .thenValue([&blk_id](auto&& err) -> BlobManager::AsyncResult< blob_id_t > {
+                                 // TODO: do we need to update repl_dev metrics?
+                                 if (err) {
+                                     LOGE("Failed to write shard info to blk_id: {}", blk_id.to_string());
+                                     return folly::makeUnexpected(BlobError(BlobErrorCode::REPLICATION_ERROR));
+                                 }
+                                 LOGD("Shard info written to blk_id:{}", blk_id.to_string());
+                                 return 0;
+                             })
+                             .get();
+        if (ret) {
+            LOGE("Failed to write shard info of blob_id {} to blk_id:{}", blob->blob_id(), blk_id.to_string());
+            free_allocated_blks();
+            return WRITE_DATA_ERR;
+        }
+
+        // Add local blob info to index & PG
+        bool success = home_obj_.local_add_blob_info(pg_id_, BlobInfo{shard_cursor_, blob->blob_id(), {0, 0, 0}});
+        if (!success) {
+            LOGE("Failed to add blob info for blob_id:{}", blob->blob_id());
+            free_allocated_blks();
+            return ADD_BLOB_INDEX_ERR;
+        }
+
+        blob_cursor_ = blob->blob_id();
```

Comment:
What is the purpose of this `blob_cursor_`?

Reply:
Previously thought it might be useful for progress tracking or resuming... but doesn't really seem to be.
Let me remove it for now.